README.md

```diff
@@ -15,7 +94,53 @@ have pre-slotted error codes for each library. The currently allocated error ran
 | [0, 0x0400) | aws-c-common |
 | [0x0400, 0x0800) | aws-c-io |
 | [0x0800, 0x1000) | aws-c-http |
-| [0x1000, 0x2000) |aws-c-amazon-flow | 
+| [0x1000, 0x2000) |aws-c-eventstream | 
 | [0x2000, 0x2800) | (reserved for future project) |
 
-Each library should begin its error codes at the beginning of its range and follow in sequence (don't skip codes).
+Each library should begin its error codes at the beginning of its range and follow in sequence (don't skip codes). Upon
+adding an AWS maintained library, an error code range must be approved and added to the above table.
+
+### Testing
+We have a high bar for test coverage, and PRs fixing bugs or introducing new functionality need to have tests before 
+they will be accepted. A couple of tips:
+
+#### Aws Test Harness
+We provide a test harness for writing unit tests. This includes an allocator that will fail your test if you have any 
+memory leaks, as well as some `ASSERT` macros. To write a test:
+
+* Create a *.c test file in the tests directory of the project. 
+* Implement one or more tests with the signature `int test_case_name(struct aws_allocator *, void *ctx)`
+* Use the `AWS_TEST_CASE` macro to declare the test.
+* Include your test in the `tests/main.c` file.
+* Include yur test in the `tests/CMakeLists.txt` file. 
+
+### Coding Style
+* No Tabs
+* Indent is 4 spaces
+* K & R style for braces
+* Space after if, before the `(`
+* Avoid C99 features in header files
+* Avoid C++ style comments e.g. `//`
+* All public API functions need C++ guards and Windows dll semantics
+* Use Unix line endings
+* SNAKE_UPPER_CASE constants, macros, and enum members.
+* snake_lower_case everything else.
+* do not use typedef struct idiom.
```

Comment:
I don't understand why we put the kibosh on this. If we're typedeffing function pointers, it also seems to make sense to me to typedef struct pointers when we want users to handle them but we don't want users to be able to declare their own instances of the structs. (i.e., opaque pointers with forward declaration of structs, so that we can change their implementation later)

Reply:
typedef structs is broadly seen as problematic. I'm not opposed to it in the ideal, but I do want to standardize on one or the other. 

I also don't mind amending this standard when we have a good use case for using it.

Reply:
I've never heard of typedef struct as being problematic, and I'm wondering what the case for that is. That being said, what I am advocating for is slightly different, i.e., typedeffing a pointer to a struct, not the struct itself. I can understand the argument that typedef struct might obscure the actual type detail to a programmer who needs to know it, but the case I am thinking about is specifically the one where we want our users *not* to know the details of the struct, i.e., when we return or modify a struct in certain API calls that we give them, and they have to in turn send those structs to future API calls. In those cases, we want to reserve the possibility to change our implementation of the struct later, so we don't want them ever dereferencing the pointer and doing something to the memory underneath. By typedeffing, we encourage our users not to think about the object as a pointer type, but simply as a thing they need to pass around from one function to another as a whole. So I might amend the wording to "Only use typedef for struct pointers, not structs themselves, when they are being forward declared in end user header files."

Reply:
http://yarchive.net/comp/linux/typedefs.html